src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/MetadataService.kt

```diff
@@ -0,0 +1,189 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement
+
+import com.amazon.opendistroforelasticsearch.indexmanagement.IndexManagementIndices
+import com.amazon.opendistroforelasticsearch.indexmanagement.elasticapi.retry
+import com.amazon.opendistroforelasticsearch.indexmanagement.elasticapi.suspendUntil
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.elasticapi.getManagedIndexMetadata
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.transport.action.updateindexmetadata.UpdateManagedIndexMetaDataAction
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.transport.action.updateindexmetadata.UpdateManagedIndexMetaDataRequest
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.util.managedIndexMetadataIndexRequest
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.util.revertISMMetadataID
+import com.amazon.opendistroforelasticsearch.indexmanagement.util.OpenForTesting
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.ExceptionsHelper
+import org.elasticsearch.action.DocWriteRequest
+import org.elasticsearch.action.bulk.BackoffPolicy
+import org.elasticsearch.action.bulk.BulkItemResponse
+import org.elasticsearch.action.bulk.BulkRequest
+import org.elasticsearch.action.bulk.BulkResponse
+import org.elasticsearch.action.support.master.AcknowledgedResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.block.ClusterBlockException
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.common.unit.TimeValue
+import org.elasticsearch.index.Index
+import org.elasticsearch.rest.RestStatus
+import java.lang.Exception
+
+/**
+ * When all nodes have same version IM plugin, MetadataService starts to move
+ * metadata from cluster state to metadata index
+ */
+@OpenForTesting
+class MetadataService(
```

Comment:
We should add a cluster setting to enable/disable this, just in case its stopped for some reason and someone needs to have it run again.

Reply:
Might go in the coordinator as thats where its scheduled

Reply:
added a metadata service enable setting, when it's updated, will disable/enable metadata service accordingly.

Normally metadata service is started same as our background sweep process, which is afterStart()/onMaster()/enable()